source/common/http/filter/gzip_filter.cc

```diff
@@ -0,0 +1,209 @@
+
+FilterHeadersStatus GzipFilter::decodeHeaders(HeaderMap& headers, bool) {
+  // TODO(gsagula): The current implementation checks for the presence of 'gzip' and if the same
+  // is followed by Qvalue. Since gzip is the only available content-encoding in Envoy at the
+  // moment, order/priority of preferred server encodings is disregarded (RFC2616-14.3).
+  skip_compression_ =
+      !(headers.AcceptEncoding() &&
+        std::regex_search(headers.AcceptEncoding()->value().c_str(), acceptEncodingRegex()));
+  return FilterHeadersStatus::Continue;
+}
+
+FilterHeadersStatus GzipFilter::encodeHeaders(HeaderMap& headers, bool end_stream) {
+  if (!end_stream && !skip_compression_ && isMinimumContentLength(headers) &&
+      isContentTypeAllowed(headers) && isCacheControlAllowed(headers) && isEtagAllowed(headers) &&
+      isLastModifiedAllowed(headers) && isTransferEncodingAllowed(headers) &&
+      !headers.ContentEncoding()) {
+    // TODO(gsagula): tests needed.
+    config_->disableVary()
+        ? headers.removeVary()
+        : headers.insertVary().value(Http::Headers::get().VaryValues.AcceptEncoding);
+
+    headers.removeContentLength();
+    headers.insertContentEncoding().value(Http::Headers::get().ContentEncodingValues.Gzip);
+    compressor_.init(config_->compressionLevel(), config_->compressionStrategy(),
+                     config_->windowBits(), config_->memoryLevel());
+  } else {
+    skip_compression_ = true;
+  }
+
+  return Http::FilterHeadersStatus::Continue;
+}
+
+FilterDataStatus GzipFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_compression_) {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  const uint64_t n_data{data.length()};
+
+  if (n_data) {
+    compressor_.compress(data, compressed_data_);
+  }
+
+  if (end_stream) {
+    compressor_.flush(compressed_data_);
+  }
+
+  if (compressed_data_.length()) {
+    data.drain(n_data);
+    data.move(compressed_data_);
+    return Http::FilterDataStatus::Continue;
+  }
+
+  return Http::FilterDataStatus::StopIterationNoBuffer;
+}
+
+bool GzipFilter::isMinimumContentLength(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_length = headers.ContentLength();
+  if (content_length) {
+    uint64_t length{};
+    StringUtil::atoul(content_length->value().c_str(), length);
+    return length >= config_->minimumLength();
+  }
+
+  const Http::HeaderEntry* transfer_encoding = headers.TransferEncoding();
+  return (transfer_encoding && transfer_encoding->value().find(
+                                   Http::Headers::get().TransferEncodingValues.Chunked.c_str()));
+}
+
+bool GzipFilter::isContentTypeAllowed(HeaderMap& headers) const {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type && !config_->contentTypeValues().empty()) {
+    std::string content_type_str{content_type->value().c_str()};
+    StringUtil::eraseFrom(content_type_str, ";");
+    StringUtil::trim(content_type_str);
```

Comment:
this is data-path critical code, hence my feeling that you should try to stay in a string_view world rather than copying the string several times here.

Reply:
Also note that you can tokenize without making any string copies.  I think you can use absl::StrSplit (https://github.com/abseil/abseil-cpp/blob/master/absl/strings/str_split.h) though it's annoying that the examlpe they give returns a vector strings.  It should be possible to return std::vector<absl::string_view> which is valid as long as the passed-in string_view is valid.